// formik-semantic-ui/dist/index.es.js
import React, { Component } from 'react';
import { connect, Field, Formik } from 'formik';
import { Button, Form, Dropdown, TextArea, Checkbox, Input, Radio } from 'semantic-ui-react';

function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var reactIs_production_min = createCommonjsModule(function (module, exports) {
Object.defineProperty(exports,"__esModule",{value:!0});var b="function"===typeof Symbol&&Symbol.for,c=b?Symbol.for("react.element"):60103,d=b?Symbol.for("react.portal"):60106,e=b?Symbol.for("react.fragment"):60107,f=b?Symbol.for("react.strict_mode"):60108,g=b?Symbol.for("react.profiler"):60114,h=b?Symbol.for("react.provider"):60109,k=b?Symbol.for("react.context"):60110,l=b?Symbol.for("react.async_mode"):60111,m=b?Symbol.for("react.forward_ref"):60112,n=b?Symbol.for("react.timeout"):60113;
function q(a){if("object"===typeof a&&null!==a){var p=a.$$typeof;switch(p){case c:switch(a=a.type,a){case l:case e:case g:case f:return a;default:switch(a=a&&a.$$typeof,a){case k:case m:case h:return a;default:return p}}case d:return p}}}exports.typeOf=q;exports.AsyncMode=l;exports.ContextConsumer=k;exports.ContextProvider=h;exports.Element=c;exports.ForwardRef=m;exports.Fragment=e;exports.Profiler=g;exports.Portal=d;exports.StrictMode=f;
exports.isValidElementType=function(a){return "string"===typeof a||"function"===typeof a||a===e||a===l||a===g||a===f||a===n||"object"===typeof a&&null!==a&&(a.$$typeof===h||a.$$typeof===k||a.$$typeof===m)};exports.isAsyncMode=function(a){return q(a)===l};exports.isContextConsumer=function(a){return q(a)===k};exports.isContextProvider=function(a){return q(a)===h};exports.isElement=function(a){return "object"===typeof a&&null!==a&&a.$$typeof===c};exports.isForwardRef=function(a){return q(a)===m};
exports.isFragment=function(a){return q(a)===e};exports.isProfiler=function(a){return q(a)===g};exports.isPortal=function(a){return q(a)===d};exports.isStrictMode=function(a){return q(a)===f};
});

unwrapExports(reactIs_production_min);
var reactIs_production_min_1 = reactIs_production_min.typeOf;
var reactIs_production_min_2 = reactIs_production_min.AsyncMode;
var reactIs_production_min_3 = reactIs_production_min.ContextConsumer;
var reactIs_production_min_4 = reactIs_production_min.ContextProvider;
var reactIs_production_min_5 = reactIs_production_min.Element;
var reactIs_production_min_6 = reactIs_production_min.ForwardRef;
var reactIs_production_min_7 = reactIs_production_min.Fragment;
var reactIs_production_min_8 = reactIs_production_min.Profiler;
var reactIs_production_min_9 = reactIs_production_min.Portal;
var reactIs_production_min_10 = reactIs_production_min.StrictMode;
var reactIs_production_min_11 = reactIs_production_min.isValidElementType;
var reactIs_production_min_12 = reactIs_production_min.isAsyncMode;
var reactIs_production_min_13 = reactIs_production_min.isContextConsumer;
var reactIs_production_min_14 = reactIs_production_min.isContextProvider;
var reactIs_production_min_15 = reactIs_production_min.isElement;
var reactIs_production_min_16 = reactIs_production_min.isForwardRef;
var reactIs_production_min_17 = reactIs_production_min.isFragment;
var reactIs_production_min_18 = reactIs_production_min.isProfiler;
var reactIs_production_min_19 = reactIs_production_min.isPortal;
var reactIs_production_min_20 = reactIs_production_min.isStrictMode;

var reactIs_development = createCommonjsModule(function (module, exports) {



if (process.env.NODE_ENV !== "production") {
  (function() {

Object.defineProperty(exports, '__esModule', { value: true });

// The Symbol used to tag the ReactElement-like types. If there is no native Symbol
// nor polyfill, then a plain number is used for performance.
var hasSymbol = typeof Symbol === 'function' && Symbol.for;

var REACT_ELEMENT_TYPE = hasSymbol ? Symbol.for('react.element') : 0xeac7;
var REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca;
var REACT_FRAGMENT_TYPE = hasSymbol ? Symbol.for('react.fragment') : 0xeacb;
var REACT_STRICT_MODE_TYPE = hasSymbol ? Symbol.for('react.strict_mode') : 0xeacc;
var REACT_PROFILER_TYPE = hasSymbol ? Symbol.for('react.profiler') : 0xead2;
var REACT_PROVIDER_TYPE = hasSymbol ? Symbol.for('react.provider') : 0xeacd;
var REACT_CONTEXT_TYPE = hasSymbol ? Symbol.for('react.context') : 0xeace;
var REACT_ASYNC_MODE_TYPE = hasSymbol ? Symbol.for('react.async_mode') : 0xeacf;
var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol.for('react.forward_ref') : 0xead0;
var REACT_TIMEOUT_TYPE = hasSymbol ? Symbol.for('react.timeout') : 0xead1;

function isValidElementType(type) {
  return typeof type === 'string' || typeof type === 'function' ||
  // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
  type === REACT_FRAGMENT_TYPE || type === REACT_ASYNC_MODE_TYPE || type === REACT_PROFILER_TYPE || type === REACT_STRICT_MODE_TYPE || type === REACT_TIMEOUT_TYPE || typeof type === 'object' && type !== null && (type.$$typeof === REACT_PROVIDER_TYPE || type.$$typeof === REACT_CONTEXT_TYPE || type.$$typeof === REACT_FORWARD_REF_TYPE);
}

function typeOf(object) {
  if (typeof object === 'object' && object !== null) {
    var $$typeof = object.$$typeof;

    switch ($$typeof) {
      case REACT_ELEMENT_TYPE:
        var type = object.type;

        switch (type) {
          case REACT_ASYNC_MODE_TYPE:
          case REACT_FRAGMENT_TYPE:
          case REACT_PROFILER_TYPE:
          case REACT_STRICT_MODE_TYPE:
            return type;
          default:
            var $$typeofType = type && type.$$typeof;

            switch ($$typeofType) {
              case REACT_CONTEXT_TYPE:
              case REACT_FORWARD_REF_TYPE:
              case REACT_PROVIDER_TYPE:
                return $$typeofType;
              default:
                return $$typeof;
            }
        }
      case REACT_PORTAL_TYPE:
        return $$typeof;
    }
  }

  return undefined;
}

var AsyncMode = REACT_ASYNC_MODE_TYPE;
var ContextConsumer = REACT_CONTEXT_TYPE;
var ContextProvider = REACT_PROVIDER_TYPE;
var Element = REACT_ELEMENT_TYPE;
var ForwardRef = REACT_FORWARD_REF_TYPE;
var Fragment = REACT_FRAGMENT_TYPE;
var Profiler = REACT_PROFILER_TYPE;
var Portal = REACT_PORTAL_TYPE;
var StrictMode = REACT_STRICT_MODE_TYPE;

function isAsyncMode(object) {
  return typeOf(object) === REACT_ASYNC_MODE_TYPE;
}
function isContextConsumer(object) {
  return typeOf(object) === REACT_CONTEXT_TYPE;
}
function isContextProvider(object) {
  return typeOf(object) === REACT_PROVIDER_TYPE;
}
function isElement(object) {
  return typeof object === 'object' && object !== null && object.$$typeof === REACT_ELEMENT_TYPE;
}
function isForwardRef(object) {
  return typeOf(object) === REACT_FORWARD_REF_TYPE;
}
function isFragment(object) {
  return typeOf(object) === REACT_FRAGMENT_TYPE;
}
function isProfiler(object) {
  return typeOf(object) === REACT_PROFILER_TYPE;
}
function isPortal(object) {
  return typeOf(object) === REACT_PORTAL_TYPE;
}
function isStrictMode(object) {
  return typeOf(object) === REACT_STRICT_MODE_TYPE;
}

exports.typeOf = typeOf;
exports.AsyncMode = AsyncMode;
exports.ContextConsumer = ContextConsumer;
exports.ContextProvider = ContextProvider;
exports.Element = Element;
exports.ForwardRef = ForwardRef;
exports.Fragment = Fragment;
exports.Profiler = Profiler;
exports.Portal = Portal;
exports.StrictMode = StrictMode;
exports.isValidElementType = isValidElementType;
exports.isAsyncMode = isAsyncMode;
exports.isContextConsumer = isContextConsumer;
exports.isContextProvider = isContextProvider;
exports.isElement = isElement;
exports.isForwardRef = isForwardRef;
exports.isFragment = isFragment;
exports.isProfiler = isProfiler;
exports.isPortal = isPortal;
exports.isStrictMode = isStrictMode;
  })();
}
});

unwrapExports(reactIs_development);
var reactIs_development_1 = reactIs_development.typeOf;
var reactIs_development_2 = reactIs_development.AsyncMode;
var reactIs_development_3 = reactIs_development.ContextConsumer;
var reactIs_development_4 = reactIs_development.ContextProvider;
var reactIs_development_5 = reactIs_development.Element;
var reactIs_development_6 = reactIs_development.ForwardRef;
var reactIs_development_7 = reactIs_development.Fragment;
var reactIs_development_8 = reactIs_development.Profiler;
var reactIs_development_9 = reactIs_development.Portal;
var reactIs_development_10 = reactIs_development.StrictMode;
var reactIs_development_11 = reactIs_development.isValidElementType;
var reactIs_development_12 = reactIs_development.isAsyncMode;
var reactIs_development_13 = reactIs_development.isContextConsumer;
var reactIs_development_14 = reactIs_development.isContextProvider;
var reactIs_development_15 = reactIs_development.isElement;
var reactIs_development_16 = reactIs_development.isForwardRef;
var reactIs_development_17 = reactIs_development.isFragment;
var reactIs_development_18 = reactIs_development.isProfiler;
var reactIs_development_19 = reactIs_development.isPortal;
var reactIs_development_20 = reactIs_development.isStrictMode;

var reactIs = createCommonjsModule(function (module) {

if (process.env.NODE_ENV === 'production') {
  module.exports = reactIs_production_min;
} else {
  module.exports = reactIs_development;
}
});

var _ReactIs$ForwardRef;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Copyright 2015, Yahoo! Inc.
 * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */


var REACT_STATICS = {
    childContextTypes: true,
    contextTypes: true,
    defaultProps: true,
    displayName: true,
    getDefaultProps: true,
    getDerivedStateFromProps: true,
    mixins: true,
    propTypes: true,
    type: true
};

var KNOWN_STATICS = {
    name: true,
    length: true,
    prototype: true,
    caller: true,
    callee: true,
    arguments: true,
    arity: true
};

var TYPE_STATICS = _defineProperty({}, reactIs.ForwardRef, (_ReactIs$ForwardRef = {}, _defineProperty(_ReactIs$ForwardRef, '$$typeof', true), _defineProperty(_ReactIs$ForwardRef, 'render', true), _ReactIs$ForwardRef));

var defineProperty = Object.defineProperty;
var getOwnPropertyNames = Object.getOwnPropertyNames;
var getOwnPropertySymbols = Object.getOwnPropertySymbols;
var getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
var getPrototypeOf = Object.getPrototypeOf;
var objectPrototype = Object.prototype;

function hoistNonReactStatics(targetComponent, sourceComponent, blacklist) {
    if (typeof sourceComponent !== 'string') {
        // don't hoist over string (html) components

        if (objectPrototype) {
            var inheritedComponent = getPrototypeOf(sourceComponent);
            if (inheritedComponent && inheritedComponent !== objectPrototype) {
                hoistNonReactStatics(targetComponent, inheritedComponent, blacklist);
            }
        }

        var keys = getOwnPropertyNames(sourceComponent);

        if (getOwnPropertySymbols) {
            keys = keys.concat(getOwnPropertySymbols(sourceComponent));
        }

        var targetStatics = TYPE_STATICS[targetComponent['$$typeof']] || REACT_STATICS;
        var sourceStatics = TYPE_STATICS[sourceComponent['$$typeof']] || REACT_STATICS;

        for (var i = 0; i < keys.length; ++i) {
            var key = keys[i];
            if (!KNOWN_STATICS[key] && !(blacklist && blacklist[key]) && !(sourceStatics && sourceStatics[key]) && !(targetStatics && targetStatics[key])) {
                var descriptor = getOwnPropertyDescriptor(sourceComponent, key);
                try {
                    // Avoid failures from read-only properties
                    defineProperty(targetComponent, key, descriptor);
                } catch (e) {}
            }
        }

        return targetComponent;
    }

    return targetComponent;
}

var hoistNonReactStatics_cjs = hoistNonReactStatics;

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var objectWithoutProperties = function (obj, keys) {
  var target = {};

  for (var i in obj) {
    if (keys.indexOf(i) >= 0) continue;
    if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
    target[i] = obj[i];
  }

  return target;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var Button$1 = hoistNonReactStatics_cjs(function (props) {
  return React.createElement(Button, _extends({}, props, { type: 'button' }));
}, Button);

Button$1.Submit = function (props) {
  return React.createElement(Button, _extends({ primary: true }, props, { type: 'submit' }));
};

Button$1.Reset = connect(function (_ref) {
  var handleReset = _ref.formik.handleReset,
      props = objectWithoutProperties(_ref, ['formik']);
  return React.createElement(Button, _extends({ basic: true }, props, { type: 'button', onClick: handleReset }));
});

var FormikDropdown = function (_Component) {
  inherits(FormikDropdown, _Component);

  function FormikDropdown(props) {
    classCallCheck(this, FormikDropdown);

    var _this = possibleConstructorReturn(this, (FormikDropdown.__proto__ || Object.getPrototypeOf(FormikDropdown)).call(this, props));

    var id = props.id,
        name = props.name;

    _this.id = id || 'field_dropdown_' + name;
    return _this;
  }

  createClass(FormikDropdown, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          label = _props.label,
          options = _props.options,
          _props$inputProps = _props.inputProps,
          inputProps = _props$inputProps === undefined ? {} : _props$inputProps,
          _props$fieldProps = _props.fieldProps,
          fieldProps = _props$fieldProps === undefined ? {} : _props$fieldProps;
      var _onChange = inputProps.onChange,
          safeInputProps = objectWithoutProperties(inputProps, ['onChange']);

      return React.createElement(Field, {
        name: name,
        render: function render(_ref) {
          var field = _ref.field,
              form = _ref.form;

          var error = form.touched[name] && form.errors[name];
          return React.createElement(
            Form.Field,
            _extends({ error: !!error }, fieldProps),
            !!label && React.createElement(
              'label',
              { htmlFor: _this2.id, onClick: function onClick() {
                  return _this2._dropdown.open();
                } },
              label
            ),
            React.createElement(Dropdown, _extends({
              ref: function ref(el) {
                return _this2._dropdown = el;
              },
              id: _this2.id,
              name: name,
              options: options,
              selectOnBlur: false,
              selectOnNavigation: false,
              selection: true
            }, safeInputProps, {
              value: field.value,
              onChange: function onChange(e, _ref2) {
                var name = _ref2.name,
                    value = _ref2.value;

                form.setFieldValue(name, value, true);
                _onChange && _onChange(e, { name: name, value: value });
              }
            })),
            error && React.createElement(
              'span',
              { className: 'sui-error-message' },
              form.errors[name]
            )
          );
        }
      });
    }
  }]);
  return FormikDropdown;
}(Component);

var FormikTextArea = function (_Component) {
  inherits(FormikTextArea, _Component);

  function FormikTextArea(props) {
    classCallCheck(this, FormikTextArea);

    var _this = possibleConstructorReturn(this, (FormikTextArea.__proto__ || Object.getPrototypeOf(FormikTextArea)).call(this, props));

    var id = props.id,
        name = props.name;

    _this.id = id || 'field_textarea_' + name;
    return _this;
  }

  createClass(FormikTextArea, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          label = _props.label,
          _props$inputProps = _props.inputProps,
          inputProps = _props$inputProps === undefined ? {} : _props$inputProps,
          _props$fieldProps = _props.fieldProps,
          fieldProps = _props$fieldProps === undefined ? {} : _props$fieldProps;
      var _onChange = inputProps.onChange,
          safeInputProps = objectWithoutProperties(inputProps, ['onChange']);

      return React.createElement(Field, {
        name: name,
        render: function render(_ref) {
          var field = _ref.field,
              form = _ref.form;

          var error = form.touched[name] && form.errors[name];
          return React.createElement(
            Form.Field,
            _extends({ error: !!error }, fieldProps),
            !!label && React.createElement(
              'label',
              { htmlFor: _this2.id },
              label
            ),
            React.createElement(TextArea, _extends({
              id: _this2.id,
              name: name,
              rows: 4
            }, safeInputProps, {
              value: field.value,
              onChange: function onChange(e, _ref2) {
                var name = _ref2.name,
                    value = _ref2.value;

                form.setFieldValue(name, value, true);
                _onChange && _onChange(e, { name: name, value: value });
              }
            })),
            error && React.createElement(
              'span',
              { className: 'sui-error-message' },
              form.errors[name]
            )
          );
        }
      });
    }
  }]);
  return FormikTextArea;
}(Component);

var FormikCheckbox = function (_Component) {
  inherits(FormikCheckbox, _Component);

  function FormikCheckbox(props) {
    classCallCheck(this, FormikCheckbox);

    var _this = possibleConstructorReturn(this, (FormikCheckbox.__proto__ || Object.getPrototypeOf(FormikCheckbox)).call(this, props));

    var id = props.id,
        name = props.name;

    _this.id = id || 'field_checkbox_' + name;
    return _this;
  }

  createClass(FormikCheckbox, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          label = _props.label,
          _props$inputProps = _props.inputProps,
          inputProps = _props$inputProps === undefined ? {} : _props$inputProps,
          _props$fieldProps = _props.fieldProps,
          fieldProps = _props$fieldProps === undefined ? {} : _props$fieldProps;
      var _onChange = inputProps.onChange,
          safeInputProps = objectWithoutProperties(inputProps, ['onChange']);

      return React.createElement(Field, {
        name: name,
        render: function render(_ref) {
          var field = _ref.field,
              form = _ref.form;

          var error = form.touched[name] && form.errors[name];
          return React.createElement(
            Form.Field,
            _extends({ error: !!error }, fieldProps),
            React.createElement(Checkbox, _extends({}, safeInputProps, {
              id: _this2.id,
              label: label,
              name: name,
              checked: field.value,
              onChange: function onChange(e, _ref2) {
                var name = _ref2.name,
                    checked = _ref2.checked;

                form.setFieldValue(name, checked, true);
                _onChange && _onChange(e, { name: name, value: checked });
              }
            })),
            error && React.createElement(
              'span',
              { className: 'sui-error-message' },
              form.errors[name]
            )
          );
        }
      });
    }
  }]);
  return FormikCheckbox;
}(Component);

var FormikInput = function (_Component) {
  inherits(FormikInput, _Component);

  function FormikInput(props) {
    classCallCheck(this, FormikInput);

    var _this = possibleConstructorReturn(this, (FormikInput.__proto__ || Object.getPrototypeOf(FormikInput)).call(this, props));

    var id = props.id,
        name = props.name;

    _this.id = id || 'field_input_' + name;
    return _this;
  }

  createClass(FormikInput, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          label = _props.label,
          _props$inputProps = _props.inputProps,
          inputProps = _props$inputProps === undefined ? {} : _props$inputProps,
          _props$fieldProps = _props.fieldProps,
          fieldProps = _props$fieldProps === undefined ? {} : _props$fieldProps;
      var _onChange = inputProps.onChange,
          safeInputProps = objectWithoutProperties(inputProps, ['onChange']);

      return React.createElement(Field, {
        name: name,
        render: function render(_ref) {
          var field = _ref.field,
              form = _ref.form;

          var error = form.touched[name] && form.errors[name];
          return React.createElement(
            Form.Field,
            _extends({ error: !!error }, fieldProps),
            !!label && React.createElement(
              'label',
              { htmlFor: _this2.id },
              label
            ),
            React.createElement(Input, _extends({
              id: _this2.id,
              name: name
            }, safeInputProps, {
              value: field.value,
              onChange: function onChange(e, _ref2) {
                var name = _ref2.name,
                    value = _ref2.value;

                form.setFieldValue(name, value, true);
                _onChange && _onChange(e, { name: name, value: value });
              }
            })),
            error && React.createElement(
              'span',
              { className: 'sui-error-message' },
              form.errors[name]
            )
          );
        }
      });
    }
  }]);
  return FormikInput;
}(Component);

var componentFactory = function componentFactory(schema) {
  if (!schema) {
    return undefined;
  }

  var typeMap = {
    dropdown: FormikDropdown,
    textarea: FormikTextArea,
    checkbox: FormikCheckbox
  };

  return Object.keys(schema).map(function (name) {
    var property = schema[name];
    var Component$$1 = typeMap[property.type] || FormikInput;
    var type = property.type,
        value = property.value,
        props = objectWithoutProperties(property, ['type', 'value']);

    if (Component$$1 === FormikInput) {
      props.inputProps = props.inputProps || {};
      props.inputProps.type = type;
    }
    return React.createElement(Component$$1, _extends({ key: name, name: name }, props));
  });
};

var FormikForm = function (_React$Component) {
  inherits(FormikForm, _React$Component);

  function FormikForm() {
    var _ref;

    var _temp, _this, _ret;

    classCallCheck(this, FormikForm);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = possibleConstructorReturn(this, (_ref = FormikForm.__proto__ || Object.getPrototypeOf(FormikForm)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      schemaComponents: componentFactory(_this.props.schema)
    }, _temp), possibleConstructorReturn(_this, _ret);
  }

  createClass(FormikForm, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          className = _props.className,
          inverted = _props.inverted,
          size = _props.size,
          ignoreLoading = _props.ignoreLoading,
          children = _props.children,
          render = _props.render,
          component = _props.component,
          _props$initialValues = _props.initialValues,
          initialValues = _props$initialValues === undefined ? {} : _props$initialValues,
          formikProps = objectWithoutProperties(_props, ['className', 'inverted', 'size', 'ignoreLoading', 'children', 'render', 'component', 'initialValues']);


      var testid = formikProps['data-testid'];
      var formProps = {
        className: className,
        inverted: inverted,
        size: size
      };

      if (testid) {
        formProps['data-testid'] = testid;
      }

      if (component) {
        console.error('\n        The formik component prop is not supported in this wrapper.\n        Import the Formik control from the formik package to use this scenario.\n        ');
        throw new Error('Unsupported prop');
      }

      var _props$schema = this.props.schema,
          schema = _props$schema === undefined ? {} : _props$schema;

      var mappedValues = Object.keys(schema || initialValues || {}).reduce(function (acc, key) {
        var property = schema[key] || {};
        acc[key] = property.value || initialValues[key] || (property.type === 'checkbox' ? false : '');
        return acc;
      }, initialValues);

      return React.createElement(
        Formik,
        _extends({ initialValues: mappedValues }, formikProps),
        function (renderProps) {
          var handleSubmit = renderProps.handleSubmit,
              isSubmitting = renderProps.isSubmitting;

          return React.createElement(
            Form,
            _extends({}, formProps, {
              onSubmit: handleSubmit,
              loading: isSubmitting && !ignoreLoading
            }),
            _this2.state.schemaComponents,
            typeof children === 'function' || typeof render === 'function' ? (render || children)(renderProps) : children
          );
        }
      );
    }
  }]);
  return FormikForm;
}(React.Component);

var Form$1 = hoistNonReactStatics_cjs(FormikForm, Form);
Form$1.Children = React.Fragment;

var FormikCheckbox$1 = function (_Component) {
  inherits(FormikCheckbox, _Component);

  function FormikCheckbox(props) {
    classCallCheck(this, FormikCheckbox);

    var _this = possibleConstructorReturn(this, (FormikCheckbox.__proto__ || Object.getPrototypeOf(FormikCheckbox)).call(this, props));

    var id = props.id,
        name = props.name,
        value = props.value;

    _this.id = id ? id + '_' + value : 'field_radio_' + name + '_' + value;
    return _this;
  }

  createClass(FormikCheckbox, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          name = _props.name,
          label = _props.label,
          value = _props.value,
          _props$inputProps = _props.inputProps,
          inputProps = _props$inputProps === undefined ? {} : _props$inputProps,
          _props$fieldProps = _props.fieldProps,
          fieldProps = _props$fieldProps === undefined ? {} : _props$fieldProps;
      var _onChange = inputProps.onChange,
          safeInputProps = objectWithoutProperties(inputProps, ['onChange']);

      return React.createElement(Field, {
        name: name,
        render: function render(_ref) {
          var field = _ref.field,
              form = _ref.form;

          var error = form.touched[name] && form.errors[name];
          return React.createElement(
            Form.Field,
            _extends({ error: !!error }, fieldProps),
            React.createElement(Radio, _extends({}, safeInputProps, {
              id: _this2.id,
              label: label,
              name: name,
              value: value,
              checked: field.value === value,
              onChange: function onChange(e, _ref2) {
                var name = _ref2.name,
                    value = _ref2.value;

                form.setFieldValue(name, value, true);
                _onChange && _onChange(e, { name: name, value: value });
              }
            })),
            error && React.createElement(
              'span',
              { className: 'sui-error-message' },
              form.errors[name]
            )
          );
        }
      });
    }
  }]);
  return FormikCheckbox;
}(Component);

// export { default as DatePicker } from "./DatePicker";
// export { default as FileUpload } from "./FileUpload";

export { Button$1 as Button, Form$1 as Form, FormikInput as Input, FormikCheckbox as Checkbox, FormikCheckbox$1 as Radio, FormikDropdown as Dropdown, FormikTextArea as TextArea };
